### ISSUE TITLE
cat $file 2>/dev/null and <$file 2>/dev/null are not equivalent if file does not exist

### ISSUE THREAD

**Issue by KaspervdHeijden** [koalaman/shellcheck#2887]
#### For bugs
- Rule Id (if any, e.g. SC1000):  SC2002 
- My shellcheck version (`shellcheck --version` or "online"):  0.9.0
- [ x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x ] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ x] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [x ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/urs/bin/env sh

file="somefile.txt"

cat "${file}" 2>/dev/null | tr '\n' ' '
```

#### Here's what shellcheck currently says:

cat "${file}" 2>/dev/null | tr '\n' ' '
    ^-- [SC2002](https://www.shellcheck.net/wiki/SC2002) (style): Useless cat. Consider 'cmd < file | ..' or 'cmd file | ..' instead.

#### Here's what I wanted or expected to see:
Nothing

- `tr` does not accept a file name argument; so the `cmd "${file}"` is not possible
- `cat "${file}" 2>/dev/null` and `<"${file}" 2>/dev/null`  **are not equivalent**; the former does _not_ output an errormessage when the file does not exist, whilst the latter does because it's the shell complaining, not an external tool.

**Comment by slycordinator** [comment]
You can suppress the output of `<"${file}"` for a non-existent file.

Depending on your needs, you could use:
`{ <"${file}"; } 2>/dev/null`
or
`( <"${file}"; ) 2>/dev/null`

**Comment by KaspervdHeijden** [comment]
Nice trick! Today I learned. Would it be possible to let shellcheck suggest this in this specific situation?

For what it's worth, I guess we're now in the territory that it's just easier to use `cat` anyway, making it a false positive (we are still creating a subshell). But that's just my option.

**Comment by slycordinator** [comment]
"we are still creating a subshell"

The option with {} doesn't create a stubshell.

**Comment by KaspervdHeijden** [comment]
Note: If I use `{ <"${file}"; } 2>/dev/null` shellcheck still does not like it:

```
^-- SC2188 (warning): This redirection doesn't have a command. Move to its command (or use 'true' as no-op).
```

**Comment by ch604** [comment]
To support our long-standing UUOC traditions, I'd like to mention that there are other cases where such a pipeline cannot be avoided with an input redirect to get identical output, at least not to my knowledge. For instance, `wc`:

```
file="name_of_file"
wc -l "$file"
# 17 name_of_file
```
In this case, we just want the line count to tell the end user. The output is going through `column`, so we need a single field here, even if it's a zero. Certainly there are plenty of ways to adjust the output with cut or sed or awk, but the lightest program which comes to mind first is `cat`:
```
cat "$file" | wc -l
# 17
```
Shellcheck does not like this (SC2002), and using input redirection here will work, _unless the file does not exist_. In this case, the line count of the file should be reported as zero. Here I have suppressed error messages:
```
wc -l 2> /dev/null < "$file"
# 17
rm "$file"
wc -l 2> /dev/null < "$file" 
# (null output)
wc -l "$file" 2> /dev/null
# (null output)
cat "$file" 2> /dev/null | wc -l
# 0 (wc -l counts zero lines of input)
```
We can try a subshell, which pipes its output into `wc`, though this is not much different to cat in the first place, and shellcheck again does not quite like this input to no program (SC2188):
```
{ < "$file" ; } 2> /dev/null | wc -l
# 0 (again, counting 0 lines of input)
```
...or, using a more complicated `test` function, which defeats the purpose of saving precious characters and clock cycles:
```
if [ ! -e "$file" ]; then echo "0"; else wc -l < "$file"; fi
# 0
```
The closest analog I've crafted is to simply account for the case:
```
wc -l 2> /dev/null < "$file" || echo "0"
# 0
```
... or just `touch` the file in the first place before attempting to fill it with my error messages. I suppose the aim of this comment is to ask to outline alternatives with identical output under the exceptions section in the wiki, which would follow a better shellcheck standard. For now, I'll be using the final `||` alternative.

**Comment by slycordinator** [comment]
[We can try a subshell, which pipes its output into wc, though this is not
much different to cat in the first place, and shellcheck again does not
quite like this input to no program (SC2188):

{ < "$file" ; } 2> /dev/null | wc -l
# 0 (again, counting 0 lines of input)]

This doesn't create a subshell.

If you want to do the same where a subshell is created, it would be..

( < "$file" ; ) 2> /dev/null | wc -l
